keps/sig-api-machinery/2855-automated-stored-versions/README.md

```diff
@@ -0,0 +1,567 @@
+
+This KEP proposes a solution that would allow the [kube-storage-version-migrator] to safely modify a CRD's `status.storedVersions` after a successful migration.
+
+[kube-storage-version-migrator]: https://github.com/kubernetes-sigs/kube-storage-version-migrator
+
+## Motivation
+
+It is desirable for the [kube-storage-version-migrator] to modify a CRD's `status.storedVersions` array after a successful migration rather than relying on a cluster admin to manually perform the same steps.
+
+### Goals
+
+- Allow the [kube-storage-version-migrator] to update a CRD's `status.storedVersions`  with the list of stored versions after a successful migration.
+- Ensure that the kube-storage-version-migrator has successfully done its job before removing a stored version from the status of the CRD.
+
+### Non-Goals
+
+- Modify a CRD's `status.storedVersions` field if the migration is not successful.
+
+## Proposal
+
+In order to safely modify a CRD's `status.storedVersions` field, the [kube-storage-version-migrator] would need to prevent changes to the CRD during a migration. CRD's could be marked as migrating using an annotation and changes to the CRD could be prevented by means of an [Validating Admission Webhook](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/).
+
+The validating admission webhook would need to:
+
+- Intercept all Update and Patch operations against CRDs that include the `ksvm.k8s.io/migrating=true` annotation.
+- Allow `/status` requests from the [kube-storage-version-migrator] service account.
+- Allow setting or removing the migrating annotation by the [kube-storage-version-migrator] service account.
+- Prevent other changes to the CRD.
+
+A summary of the new migration workflow can be found below:
+
+1. The [kube-storage-version-migrator] notices a new stored version for the `foo` CRD.
+2. The  `ksvm.k8s.io/migrating=true`  annotation is applied to the`foo` CRD that defines the resource being migrated.
+3. The `StorageVersionMigration` CR is created for the resource introduced by the `foo` CRD.
+4. The [kube-storage-version-migrator] begins migrating all `foo` resources as it does today.
+5. When a migration is successful, the `foo` CRD's `status.storedVersion` is patched to only include the new stored version.
+6. The `ksvm.k8s.io/migrating=true` annotation is removed from the `foo` CRD.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster admin, I would like the [kube-storage-version-migrator] to automatically update a CRD's `status.storeVersions` after a successful a migration.
+
+### Notes/Constraints/Caveats (Optional)
+
+- The introduction of an admission webhook would require a valid CA, possibly managed by [cert-manager](https://github.com/jetstack/cert-manager). At a minimum, documentation should be provided that outlines how to provide a cert for the webhook.
```

Comment:
cert-manager uses CRDs. Does this cause a cyclic dependency?

Reply:
Cert-manager can be deployed on clusters without ksvm.

If the proposed changes are introduced, ksvm could not run on cluster without some entity providing valid certificates for the admission webhook. To your point, when a cluster administrator relies on an API introduced by a CRD to provision these certs (such as the case with cert-manager) ksvm then begins to manage the storage versions of CRD introduced by the cert provisioner and relies on the cert provisioner for cert management.

With that in mind, the only instance I see where we could hit an issue is when:
- ksvm relies on a cert-manager that uses CRDs (in this case cert manager)
- A migration for one of the CRDs owned by the cert-manager is started
- A new version of cert manager is deployed that is not compatible with the CRD being migrated and introduces a new stored version.
- The cert-manager upgrade fails because of the admission webhook. Cert-manager is no longer able to provision new certs (key failure)
- The certificate for ksvm expires (key failure)
- Changes to the CRD are now blocked indefinitely without a cluster admin deleting the admission webhook, stopping the migration, and finishing the upgrade / rolling back to the previous version of the cert-manager.

This seems like an edge case that we should consider, but it can still be manually resolved. Are there other concerns that you have?

Reply:
Going to add this under the `Risks and Mitigations` section.

Reply:
Called out the cyclical dependency in the KEP.